[autograd Function] Add private API to not materialize grads for non-differentiable outputs

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104787
* __->__ #104291

Fixes https://github.com/pytorch/pytorch/issues/104272

This PR adds a new private API `materialize_non_diff_grads` (default True) such that when set to False, grad outputs corresponding to outputs marked non-differentiable would receive None instead of a zero-filled tensor. This is overrides the setting of `materialize_grads`, i.e. grad outputs corresponding non-differentiable outputs would still be None even if `materialize_grads=True` (the default). 

cc @ezyang @gchanan

soulitzer committed:
[autograd Function] Do not materialize grads for non-differentiable outputs

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/function.h:
```diff
@@ -521,6 +521,11 @@ struct TORCH_API Node : std::enable_shared_from_this<Node> {
     return retains_grad_hooks_;
   }
 
+  virtual void set_non_differentiable_idx(
```
This feels not right.
I feel like this shouldn't be on the top level Node class.
I like the idea of doing this setting in the shared code but since you need special handling in PyNode and CppNode anyways, I feel like the apply code for each of these could update these fields directly.

soulitzer replied:
Yeah thats fair, let me try something else.

soulitzer replied:
Using a new approach now that still does things in the shared code, but no longer places methods on the top level Node class.
The new sin we're committing here though is that we're overloading the default constructed InputMetadata to mean something special for custom functions, but maybe it isn't that bad.

soulitzer committed:
Update on "[autograd Function] Do not materialize grads for non-differentiable outputs"


Fixes https://github.com/pytorch/pytorch/issues/104272

cc ezyang gchanan

[ghstack-poisoned]

soulitzer committed:
Update on "[autograd Function] Do not materialize grads for non-differentiable outputs"


Fixes https://github.com/pytorch/pytorch/issues/104272

cc ezyang gchanan

[ghstack-poisoned]

soulitzer commented:
There's still a functorch failure `test_vmapvjpvmap_SortGenVmapAutogradFunction` I'll have to sort out later, but curious what you think of this new approach @albanD

soulitzer committed:
Update on "[autograd Function] Do not materialize grads for non-differentiable outputs"


Fixes https://github.com/pytorch/pytorch/issues/104272

cc ezyang gchanan

[ghstack-poisoned]

soulitzer commented on torch/_functorch/vmap.py:
```diff
@@ -125,7 +125,7 @@ def _process_batched_inputs(
         flat_in_dims: List[Any], flat_args: List[Any], vmap_level: int, args_spec) -> Tuple:
     # See NOTE [Ignored _remove_batch_dim, _add_batch_dim]
-    batched_inputs = [arg if in_dim is None else
+    batched_inputs = [arg if (in_dim is None or arg is None) else
```
This might be wrong, maybe we should handle this at the autograd Function level

soulitzer replied:
@albanD @zou3519 
I'm not sure how hard it is to add support for this at the functorch compatible autograd Function level. From looking at the ops in autograd_function_db it looks like we currently assume all grad_outputs to be non-none? The only op in autograd_function_db that takes in a None for its backward is xfailed everywhere (NumpyCubeNotComposableAutogradFunction), and we also don't support set_materialize_grads either.

I can think more about it if needed, but if it is non-trivial to add support for functorch, I can imagine it being difficult for certain other users as well. How do we feel about just adding a new config to opt-out restore back to materialization of grad_outputs?

zou3519 replied:
We shouldn't change the vmap behavior. We will need to think more about what the vmap behavior on None is (it involves pytrees).

A simple fix is to modify `restore_vmap` to handle this situation. The implication is also that `generate_vmap_rule=True` doesn't handle `ctx.set_materialize_grads(False)` correctly (looking through autograd_function_db, it looks like we didn't handle it).

Though we might want to be careful around changing the default:
- Many users will assume that the grad_outputs are always Tensors. This would break them.
- One potential way around this is to add a new ctx config. I'd be fine with that.
- Another option is to add this behavior to ctx.set_materialize_grads ? I would have expected ctx.set_materialize_grads to already control this behavior, but it sounds like it doesn't

vadimkantorov replied:
could it instead make these grads into meta tensors or sth like that? so that they are not None, but not materialized? or EfficientZeroTensors? or just a scalar zero broadcasted to the target shape? these should not take much space

albanD commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -342,7 +342,8 @@ variable_list CppNode<T>::apply(variable_list&& inputs) {
   backward_inputs.reserve(num_inputs);
   for (const auto i : c10::irange(num_inputs)) {
-    if (inputs[i].defined() || !ctx_.materialize_grads_) {
+    if (inputs[i].defined() || !ctx_.materialize_grads_ ||
```
Shouldn't these still be undefined even when materialize_grad is True ?

soulitzer replied:
Actually no, I don't think we want to do this now. 

Yes, I don't imagine anyone actually needing zero-filled grads for non-differentiable outputs, but I think this can break some people as Richard points out above - we also have quite a few failures in tests.

So for the purposes of making this minimally BC-breaking, I prefer to make this a opt-in mechanism instead, as in
I do want to materialize when ctx.set_materialize_grads(False).

My reasoning is that:
- it is rare for people who can't handle None to be using set_materialize_grads(False)
- people care about perf will find out about using set_materialize_grads(False)

albanD replied:
Sounds ok to me!

albanD reviewed (commented):
Looks pretty cool!
nit on the force materialize.

I'm not sure about the vmap issue though and I'll let @zou3519 answer.

soulitzer committed:
Update on "[autograd Function] Do not materialize grads for non-differentiable outputs"


Fixes https://github.com/pytorch/pytorch/issues/104272

cc ezyang gchanan

[ghstack-poisoned]

soulitzer committed:
Update on "[autograd Function] Do not materialize grads for non-differentiable outputs"


Fixes https://github.com/pytorch/pytorch/issues/104272

cc ezyang gchanan

[ghstack-poisoned]

soulitzer committed:
Update on "[autograd Function] Do not materialize grads for non-differentiable outputs"


Fixes https://github.com/pytorch/pytorch/issues/104272

cc ezyang gchanan

[ghstack-poisoned]

soulitzer committed:
Update on "[autograd Function] Do not materialize grads for non-differentiable outputs"


Fixes https://github.com/pytorch/pytorch/issues/104272

cc ezyang gchanan

[ghstack-poisoned]

soulitzer commented:
I realize I completely misunderstood the original requirements - The ask in https://github.com/pytorch/pytorch/pull/98659#pullrequestreview-1376822560 is specifically that there is a way to not materialize grad outs corresponding to non-diff outputs even if set_materialize_grads=True. 

>  I would have expected ctx.set_materialize_grads to already control this behavior, but it sounds like it doesn't

And those expectations would've been correct actually. What I outline here is already the current behavior! (I didn't bother to check for some reason) https://github.com/pytorch/pytorch/pull/104291#discussion_r1245701279

Anyway, due to bc-breaking concerns, I've updated this PR to add a separate private setting that does this rather than changing the default behavior. It's private because I don't expect users to find this useful in general; they should be using set_materialize_grads instead.

ezyang reviewed (commented):
Don't forget the pyi annotation for the new binding

soulitzer committed:
Update on "[autograd Function] Add private API to not materialize grads for non-differentiable outputs"


Fixes https://github.com/pytorch/pytorch/issues/104272

cc ezyang gchanan

[ghstack-poisoned]

soulitzer committed:
Update on "[autograd Function] Add private API to not materialize grads for non-differentiable outputs"


Fixes https://github.com/pytorch/pytorch/issues/104272

cc ezyang gchanan

[ghstack-poisoned]

Chillee commented:
@soulitzer Sorry, to be clear, what exactly is the resulting change this PR is going to make? Concretely, I think we need the option to have non-differentiable outputs never filled in (without setting materialize_grads=False).

soulitzer committed:
Update on "[autograd Function] Add private API to not materialize grads for non-differentiable outputs"


Fixes https://github.com/pytorch/pytorch/issues/104272

This PR adds a new private API `materialize_non_diff_grads` (default True) such that when set to False, grad outputs corresponding to outputs marked non-differentiable would receive None instead of a zero-filled tensor. This is overrides the setting of `materialize_grads`, i.e. grad outputs corresponding non-differentiable outputs would still be None even if `materialize_grads=True` (the default). 

cc ezyang gchanan

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -363,7 +363,9 @@ static optional_variable_list _process_backward_mode_ad(
   for (const auto i : c10::irange(num_outputs)) {
-    // For outputs that are not tensors, put a placeholder undefined input.
+    // For outputs that are not tensors put a placeholder undefined input.
+    // Note that below, we also put a placeholder undefined inputs for when
```
Nit: merge the two sentences: we put undefined for non-Tensor and Tensors that are not differentiable

albanD commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -1149,6 +1152,29 @@ int THPFunction_set_materialize_grads(
+    THPFunction* self,
+    void* _unused) {
+  HANDLE_TH_ERRORS
+  return self->materialize_non_diff_grads ? Py_True : Py_False;
```
You're missing refcounting!
Use the Py_RETURN_TRUE / Py_RETURN_FALSE macros

soulitzer replied:
good catch!

albanD commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -1384,6 +1410,12 @@ static struct PyGetSetDef THPFunction_properties[] = {
+     (setter)THPFunction_set_materialize_non_diff_grads,
+     nullptr,
+     nullptr},
+    {nullptr},
```
This extra nullptr is not needed?

soulitzer replied:
whoops

albanD commented on torch/csrc/autograd/python_function.h:
```diff
@@ -93,6 +93,13 @@ struct THPFunction {
   // Default is true.
   bool materialize_grads;
 
+  // boolean indicating whther to materialize output grad tensors
```
whether

soulitzer committed:
Update on "[autograd Function] Add private API to not materialize grads for non-differentiable outputs"


Fixes https://github.com/pytorch/pytorch/issues/104272

This PR adds a new private API `materialize_non_diff_grads` (default True) such that when set to False, grad outputs corresponding to outputs marked non-differentiable would receive None instead of a zero-filled tensor. This is overrides the setting of `materialize_grads`, i.e. grad outputs corresponding non-differentiable outputs would still be None even if `materialize_grads=True` (the default). 

cc ezyang gchanan

[ghstack-poisoned]

albanD reviewed (approved):
Thanks!

soulitzer commented:
@pytorchbot merge -f "Unrelated failures"
